### PR TITLE
Optimize decryptObjectInfo

### PR DESCRIPTION
**Pull request by klauspost**
## Motivation and Context

`decryptObjectInfo` is a significant bottleneck when listing objects.

Reduce the allocations for a significant speedup. This PR covers the basic optimizations.

The MB/s is the number of decrypted objects per second.

Requires:

https://github.com/minio/sio/pull/40

`go.mod` replace term until upstream is available.

```
λ benchcmp before.txt after.txt
benchmark                          old ns/op     new ns/op     delta
Benchmark_decryptObjectInfo-32     24260928      808656        -96.67%

benchmark                          old MB/s     new MB/s     speedup
Benchmark_decryptObjectInfo-32     0.04         1.24         31.00x

benchmark                          old allocs     new allocs     delta
Benchmark_decryptObjectInfo-32     75112          48996          -34.77%

benchmark                          old bytes     new bytes     delta
Benchmark_decryptObjectInfo-32     287694772     4228076       -98.53%
```

## How to test this PR?

Tests/benchmarks are included.

## Types of changes
- [x] Optimization.

### CODE REVIEW

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10726-2009bbe/gopath/src/github.com/minio
+ git -C 10726-2009bbe/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10726-2009bbe/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10726-2009bbe/gopath/src/github.com/minio/minio fetch --quiet minio pull/10726/head:pr10726
+ git -C 10726-2009bbe/gopath/src/github.com/minio/minio checkout --quiet pr10726
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10726-2009bbe/gopath
+ make -C 10726-2009bbe/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/sio@v0.2.0: parsing ../sio/go.mod: open /home/mint-auto/mint-auto/10726-2009bbe/gopath/src/github.com/minio/sio/go.mod: no such file or directory
go: github.com/minio/sio@v0.2.0: parsing ../sio/go.mod: open /home/mint-auto/mint-auto/10726-2009bbe/gopath/src/github.com/minio/sio/go.mod: no such file or directory
make: *** [build] Error 1

```

File: cmd/crypto/kms.go [hunk 1]
```diff
@@ -67,6 +67,48 @@ func (c Context) WriteTo(w io.Writer) (n int64, err error) {
 	return n + int64(nn), err
 }
 
+// AppendTo appends the context in a canonical from to dst.
+//
+// AppendTo sorts the context keys and writes the sorted
+// key-value pairs as canonical JSON object to w.
+func (c Context) AppendTo(dst []byte) (output []byte) {
+	if len(c) == 0 {
+		return append(dst, '{', '}')
```
Comment (harshavardhana) [inline cmd/crypto/kms.go:80]: Shouldn't we be using bufio or bytes.Buffer here or something here?

Comment (harshavardhana) [reply cmd/crypto/kms.go:80]: Looks very tricky to test :-)

File: cmd/crypto/kms.go [hunk 2]
```diff
@@ -67,6 +67,48 @@ func (c Context) WriteTo(w io.Writer) (n int64, err error) {
 	return n + int64(nn), err
 }
 
+// AppendTo appends the context in a canonical from to dst.
+//
+// AppendTo sorts the context keys and writes the sorted
+// key-value pairs as canonical JSON object to w.
+func (c Context) AppendTo(dst []byte) (output []byte) {
```
Comment (aead) [inline cmd/crypto/kms.go:78]: I think we can simplify the API to `func (c Context) MarshalText() ([]byte, error)` and implement [`encoding.TextMarshaler`](https://golang.org/pkg/encoding/#TextMarshaler) - (it never returns an error).

I don't think we need to append to an existing slice?!

**Review by aead — COMMENTED** [review]
LGTM, nice optimization!

File: cmd/crypto/kms.go (hunk 2 above)
Comment (klauspost) [reply cmd/crypto/kms.go:78]: It was mostly to keep it flexible. When it receives a destination we are not forcing an allocation, but leaving that to the caller.

If we should need to do 1000 we can reuse the destination. Similar if we should need to serialize it we can do it without copying the result.

I know it isn't being used, but it is a "might as well" in this case.

Comment (aead) [reply cmd/crypto/kms.go:78]: https://blog.filippo.io/efficient-go-apis-with-the-inliner/

Comment (klauspost) [reply cmd/crypto/kms.go:78]: @aead Yes. Do you have a point with that link?

Comment (aead) [reply cmd/crypto/kms.go:78]: Yes, by implementing `encoding.Marshaler` like this:
```go
func (c Context) MarshalText() ([]byte, error) {
      text := make(...)
      return c.marshalText(text), nil
}
```
you get both - a "consumer-friendly" API (+ being a `encoding.Marshaler`) + no heap allocation due to the inliner.

Comment (klauspost) [reply cmd/crypto/kms.go:78]: @aead You don't get inlining when using interfaces and values will always be seen to escape when used as an interface. The "mid-stack inliner" is unreliable and relies on an arbitrary function score.

Real world:
```
func BenchmarkContext_AppendTo(b *testing.B) {
	c := Context{"0": "1", "-": "2", ".": "#"}
	dst := make([]byte, 0, b.N*1024)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		dst = c.AppendTo(dst)
	}
}

func BenchmarkContext_MarshalText(b *testing.B) {
	c := Context{"0": "1", "-": "2", ".": "#"}
	dst := make([]byte, 0, b.N*1024)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		x, _ := c.MarshalText()
		dst = append(dst, x...)
	}
}
```

The marshal func:
```
func (c Context) MarshalText() (output []byte, err error) {
	dst := make([]byte, 0, 1024)
	if len(c) == 0 {
		return append(dst, '{', '}'), nil
	}

	// No need to copy+sort
	if len(c) == 1 {
		dst = append(dst, '{')
		for k, v := range c {
			dst = append(dst, '"')
			dst = append(dst, []byte(k)...)
			dst = append(dst, '"', ':', '"')
			dst = append(dst, []byte(v)...)
			dst = append(dst, '"')
		}
		return append(dst, '}'), nil
	}

	sortedKeys := make([]string, 0, len(c))
	for k := range c {
		sortedKeys = append(sortedKeys, k)
	}
	sort.Strings(sortedKeys)

	dst = append(dst, '{')
	for i, k := range sortedKeys {
		dst = append(dst, '"')
		dst = append(dst, []byte(k)...)
		dst = append(dst, '"', ':', '"')
		dst = append(dst, []byte(c[k])...)
		dst = append(dst, '"')
		if i < len(sortedKeys)-1 {
			dst = append(dst, ',')
		}
	}
	return append(dst, '}'), nil
}
```

Results: 
```
BenchmarkContext_AppendTo-32    	 3726744	       321 ns/op	      80 B/op	       2 allocs/op
BenchmarkContext_MarshalText-32    	 1892852	       634 ns/op	    1104 B/op	       3 allocs/op
```

The 80B/op comes from the sorting. Neither is end of the world, but I also see no reason to choose the slower one. This is not running concurrently. There the hit would be bigger.

Comment (klauspost) [reply cmd/crypto/kms.go:78]: I could add

```
func (c Context) MarshalText() (output []byte, err error) {
	return c.AppendTo(make([]byte, 0, 128)), nil
}
```
Though I don't really see a point.

**Comment by klauspost** [comment]
@aead If you can tag sio 0.2.1 I can update the `go.mod`

**Review by harshavardhana — COMMENTED** [review]
> @aead If you can tag sio 0.2.1 I can update the `go.mod`

@klauspost will add it

**Review by harshavardhana — CHANGES_REQUESTED** [review]
v0.2.1 is available

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10726-7394d0b/gopath/src/github.com/minio
+ git -C 10726-7394d0b/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10726-7394d0b/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10726-7394d0b/gopath/src/github.com/minio/minio fetch --quiet minio pull/10726/head:pr10726
+ git -C 10726-7394d0b/gopath/src/github.com/minio/minio checkout --quiet pr10726
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10726-7394d0b/gopath
+ make -C 10726-7394d0b/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/sio@v0.2.1: parsing ../sio/go.mod: open /home/mint-auto/mint-auto/10726-7394d0b/gopath/src/github.com/minio/sio/go.mod: no such file or directory
go: github.com/minio/sio@v0.2.1: parsing ../sio/go.mod: open /home/mint-auto/mint-auto/10726-7394d0b/gopath/src/github.com/minio/sio/go.mod: no such file or directory
make: *** [build] Error 1

```

File: cmd/crypto/kms.go (hunk 1 above)
Comment (klauspost) [reply cmd/crypto/kms.go:80]: @harshavardhana How so?

Comment (harshavardhana) [reply cmd/crypto/kms.go:80]: AppendTo is tricky with all the custom strings to generate JSON is not easy to read

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
> LGTM, nice optimization! 

PTAL @aead

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#77a7130c-8080-4b4c-a655-55df355b9655) |
<a name="77a7130c-8080-4b4c-a655-55df355b9655"></a>
### 10726-2b662b4/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31421
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 236da7079bc9:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 40 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 50 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 3 minutes and 33 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 38 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 46 seconds
{
  "args": {
    "bucketName": "minio-go-test-g99uvymibf3g9vds",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1030,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 20 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 79,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 18 minutes and 28 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 55 seconds
(12/15) Running s3select tests ... done in 39 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/crypto/kms.go (hunk 1 above)
Comment (klauspost) [reply cmd/crypto/kms.go:80]: @harshavardhana Let me check if it can be done without allocations.

Comment (klauspost) [reply cmd/crypto/kms.go:80]: Nice. A bytes.Buffer does not escape, so no heap allocs and nicer code.

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#b8fe0ffc-4d6d-48de-a638-45b51cf97f6d) |
<a name="b8fe0ffc-4d6d-48de-a638-45b51cf97f6d"></a>
### 10726-7143a31/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31199
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f9c837008fa9:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 48 seconds
(4/15) Running aws-sdk-ruby tests ... done in 21 seconds
(5/15) Running awscli tests ... done in 2 minutes and 50 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 28 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 36 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 57 seconds
{
  "args": {
    "bucketName": "minio-go-test-3b6idbzvvsimzuo9",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1215,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 14 seconds
(10/15) Running minio-js tests ... FAILED in 47 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 79,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 18 minutes and 4 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 23 seconds
(12/15) Running s3select tests ... done in 38 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Review by aead — APPROVED** [review]

**Pull request merged**